### ISSUE TITLE
IsEquivalentTo Transitive Property For Dictionary Values Fails Unexpectedly

### ISSUE THREAD

**Issue by aolszowka** [nunit/nunit#3441]
NUnit 3.12.0
NUnit Runner 3.15.1

Hi,

Consider the following Toy Unit Test

```c#
public void IsEquivalentTransitiveWhenNestedInDictionary()
{
    HashSet<string> actual1 = new HashSet<string>() { "B", "C" };
    HashSet<string> expected1 = new HashSet<string>() { "B", "C" };

    Assert.That(actual1, Is.EquivalentTo(expected1), "Test Case 1 Failed");

    Dictionary<string, HashSet<string>> actual2 = new Dictionary<string, HashSet<string>> { { "A", new HashSet<string>() { "B", "C" } } };
    Dictionary<string, HashSet<string>> expected2 = new Dictionary<string, HashSet<string>> { { "A", new HashSet<string>() { "B", "C" } } };

    Assert.That(actual2, Is.EquivalentTo(expected2), "Test Case 2 Failed");

    HashSet<string> actual3 = new HashSet<string>() { "B", "C" };
    HashSet<string> expected3 = new HashSet<string>() { "C", "B" };

    Assert.That(actual3, Is.EquivalentTo(expected3), "Test Case 3 Failed");

    Dictionary<string, HashSet<string>> actual4 = new Dictionary<string, HashSet<string>> { { "A", new HashSet<string>() { "B", "C" } } };
    Dictionary<string, HashSet<string>> expected4 = new Dictionary<string, HashSet<string>> { { "A", new HashSet<string>() { "C", "B" } } };

    Assert.That(actual4, Is.EquivalentTo(expected4), "Test Case 4 Failed");
}
```

Test Case 4 fails; is this expected behavior?

Possibly Related: #3073 ?

**Comment by CharliePoole** [comment]
This is not behavior I would expect.

NUnit's notion of Equivalence is that it applies to two collections and is satisfied if the two collections contain the same elements, in any order. To test whether two elements are the same, NUnit uses Equality, not Equivalence. So, Equivalence IMO is not supposed to be transitive.

__However__, it's also my opinion that two `HashSets` containing the same values ought to be equal, since set equality does not have anything to do with order.

Various other team members have worked on this since I first defined Equivalence for collections, so lets see what others think.

BTW... it's also my view that Equivalence and Equality for two dictionaries should be the same. They are different only for collection where order matters.

**Comment by mikkelbu** [comment]
This has not much to do with dictionaries, but rather nested "collections", as far as I can tell. The following also fails
```csharp
[Test]
public void IsEquivalentTransitiveWhenNestedInDictionary5()
{
    var actual5 = new List<HashSet<string>> { { new HashSet<string>() { "B", "C" } } };
    var expected5 = new List<HashSet<string>> { { new HashSet<string>() { "C", "B" } } };

    Assert.That(actual5, Is.EquivalentTo(expected5), "Test Case 5 Failed");
}
```
The problem is that comparing the outer "collections" we end up in `src\NUnitFramework\framework\Constraints\CollectionTally.cs` where we try to match the individual elements, and this matching is done using `CollectionTally.ItemsEqual(object expected, object actual)`.
So we are matching the elements using `NUnitEqualityComparer.AreEqual` and since we don't have any special handling of `HashSet` we end up using the last comparer in `src\NUnitFramework\framework\Constraints\NUnitEqualityComparer.cs` which is comparing the two `HashSet`s as `IEnumerable`s, hence using `src\NUnitFramework\framework\Constraints\Comparers\EnumerablesComparer.cs` and here order matters.

So I think the problem is that we don't have any special handling of `HashSet` and end up comparing the `HashSet`s as `IEnumerable`s.

**Comment by aolszowka** [comment]
Please feel free to change the title of this to better reflect the underlying cause; I may have used the word "transitive" in an inappropriate way (I am still not quite sure how to describe this property).

FWIW in my case I actually converted to a SortedSet<string> which was closer to what I needed anyway, and because it was sorted behaved as expected.

**Comment by CharliePoole** [comment]
@aolszowka Just in case, I'll mention that `EqualTo` is usually better to use for collections __unless__ you are trying to disregard order. Some folks are not aware that NUnit defines collection equality as it does and use the less efficient `EquivalentTo` instead.

**Comment by aolszowka** [comment]
@CharliePoole Yup; I remember reading that somewhere (probably a post from you at some point).

I was intentionally using EquivalentTo because of my original use of `HashSet<string>` which IIRC does not guarantee any type of ordering if you simply enumerated though it.

**Comment by OsirisTerje** [comment]
This is still not working with the latest versions.
Added repro here:  https://github.com/nunit/nunit.issues/tree/main/Issue3441 

@stevenaw It seems CollectionTally is involved here.  Should we take another look there wrt this issue?

**Comment by manfred-brands** [comment]
@OsirisTerje it is the nested collection that is the issue. 
As @mikkelbu said, we need do detect a IReadOnlySet and use EquivalentTo

**Comment by OsirisTerje** [comment]
Yes, absolutely.   

> The problem is that comparing the outer "collections" we end up in src\NUnitFramework\framework\Constraints\CollectionTally.cs where we try to match the individual elements, and this matching is done using CollectionTally.ItemsEqual(object expected, object actual).

And if I have got this right, we need to address this inside the CollectionTally.   So the question is,  can we do this while we're working on the CollectionTally class now ?  

And the code referred above is from the non-generic version, the new code from the Generic version doesn't have any ItemsEqual method.  I believe we have touched all this code since 2019 (Mikkels comments), including the NUnitEqualityComparer and the EnumerablesComparer.  
We would need to add something to keep the type of what we compare through this in order to compare them correctly.

I also see the code

<img width="885" height="88" alt="Image" src="https://github.com/user-attachments/assets/26de8800-0296-4894-990c-19f87243d64d" />

which doesn't quite ring right with me.   Also, we don't have any CollectionItemsEquivalentConstraint.  So it then depends on which type of comparer we send down.  Inside the CollectionItemsEqualConstraint we have hardcoded in the NUnitEqualityComparer, so it might be here we need to do something.

So my question is:  Is this something we can (or will) do now?

**Comment by CharliePoole** [comment]
@OsirisTerje At a more fundamental level, equality and equivalence of dictionaries or two hashsets should give the same result. See comments by me and @mikkelbu back in 2019. Is it possible that the introduction of generics has broken this?

**Comment by OsirisTerje** [comment]
It didn't work in 2019 when we didn't have generics - that is pretty new - so I don't think that is the reason.

> At a more fundamental level, equality and equivalence of dictionaries or two hashsets should give the same result.

Yes, agree, since both HashSets and Dictionaries are un-ordered, that makes sense.

**Comment by stevenaw** [comment]
From what I can recall, there is a place within the DictionaryComparer that we invoke the collection tally. It would've been converted to use the generic CollectionTally very recently, within the last 2-3 weeks within the PR to deprecate the non-generic version. In that sense, consistent reproducibility of this bug today is "good" in that it supports the generic version being functionally equivalent!

I suspect then, without reading the full history or having looked at the code closely, that that may be a place to look at ensuring that any equivalency assertions are carried through the nunitequalitycomparer to nested levels (especially if custom equality rules are applied, such as `IgnoreCase` etc)
